server/controllers/admin.py

```diff
@@ -434,6 +434,45 @@ def templates(cid, aid):
                            assignment=assignment, form=form, courses=courses,
                            current_course=current_course)
 
+@admin.route("/course/<int:cid>/assignments/<int:aid>/publish",
+                methods=['GET', 'POST'])
+@is_staff(course_arg='cid')
+def publish_scores(cid, aid):
+    courses, current_course = get_courses(cid)
+    assign = Assignment.query.filter_by(id=aid, course_id=cid).one_or_none()
+    if not Assignment.can(assign, current_user, 'publish'):
+        flash('Insufficient permissions', 'error')
+        return abort(401)
+    form = forms.PublishScoresWithTags()
+    if form.validate_on_submit():
+        tag = form.grades.data
+        hide = form.hide.data
+        if hide:
+            if tag not in assign.published_scores:
+                flash("{visibility} scores for {assignment} already hidden".format(
+                    visibility=tag.title(), assignment=assign.display_name), "success")
+            else:
+                assign.hide_score(tag)
+                flash("Hid {assignment} {visibility} scores".format(
+                    assignment=assign.display_name, visibility=tag), "success")
+        else:
+            if tag in assign.published_scores:
+                flash("{visibility} scores for {assignment} already published".format(
+                    visibility=tag.title(), assignment=assign.display_name), "success")
+            elif tag == 'revision' and 'composition' not in assign.published_scores:
+                flash("{visibility} scores for {assignment} cannot be published with hidden composition scores".format(
+                    visibility=tag.title(), assignment=assign.display_name), "warning")
+            else:
+                assign.publish_score(tag)
+                flash("Published {assignment} {visibility} scores".format(
+                    assignment=assign.display_name, visibility=tag), "success")
+            
```

Comment:
This should redirect to `url_for('.publish_scores', ...)`, so that if you refresh the page after you submit you won't resubmit the form. https://en.wikipedia.org/wiki/Post/Redirect/Get